src/graphcast_agent/mod.rs

```diff
@@ -71,54 +154,79 @@ pub struct GraphcastAgent {
-    /// But in this case we will override it with something much more simple.
     /// # Examples
     ///
     /// ```ignore
-    /// let agent = GraphcastAgent::new(
-    ///     String::from("1231231231231231231231231231231231231231231231231231231231231230"),
-    ///     String::from("https://goerli.infura.io/v3/api_key"),
-    ///     "test_radio",
-    ///     "https://api.thegraph.com/subgraphs/name/hopeyen/gossip-registry-test",
-    ///     "https://gateway.testnet.thegraph.com/network",
-    ///     vec![String::from("/ip4/127.0.0.1/tcp/60000/p2p/16Uiu2YAmDEieEqD5dHSG85G8H51FUKByWoZx7byMy9AbMEgjd5iz")],
-    ///     Some("test_namespace_in_pubsub_topic"),
-    ///     Some(["some_subgraph_hash"].to_vec()),
-    ///     String::from("waku_node_key_can_be_same_as_private1231231231231231231231231230"),
-    ///     Some(String::from("0.0.0.0")),
-    ///     Some(String::from("60000")),
-    ///     Some(String::from(/ip4/321.1.1.2/tcp/60001/p2p/16Uiu2YAmDEieEqD5dHSG85G8H51FUKByWoZx7byMysomeoneelse")),
-    /// )
+    /// let config = GraphcastAgentConfig {
+    ///     wallet_key: String::from("1231231231231231231231231231231231231231231231231231231231231230"),
+    ///     radio_name: "test_radio",
+    ///     registry_subgraph: String::from("https://api.thegraph.com/subgraphs/name/hopeyen/gossip-registry-test"),
+    ///     network_subgraph: String::from("https://gateway.testnet.thegraph.com/network"),
+    ///     graph_node_endpoint: String::from("https://api.thegraph.com/index-node/graphql"),
+    ///     boot_node_addresses: vec![String::from("/ip4/127.0.0.1/tcp/60000/p2p/16Uiu2YAmDEieEqD5dHSG85G8H51FUKByWoZx7byMy9AbMEgjd5iz")],
+    ///     graphcast_namespace: Some("test_namespace_in_pubsub_topic"),
+    ///     subtopics: vec![String::from("some_subgraph_hash")],
+    ///     waku_node_key: Some(String::from("waku_node_key_can_be_same_as_private1231231231231231231231231230")),
+    ///     waku_host: Some(String::from("0.0.0.0")),
+    ///     waku_port: Some(String::from("60000")),
+    ///     waku_addr: Some(String::from("/ip4/321.1.1.2/tcp/60001/p2p/16Uiu2YAmDEieEqD5dHSG85G8H51FUKByWoZx7byMysomeoneelse")),
+    /// };
+    ///
+    /// let agent = GraphcastAgent::new(config).await?;
     /// ```
-    #[allow(clippy::too_many_arguments)]
-    pub async fn new(
-        wallet_key: String,
-        radio_name: &'static str,
-        registry_subgraph: &str,
-        network_subgraph: &str,
-        graph_node_endpoint: &str,
-        boot_node_addresses: Vec<Multiaddr>,
-        graphcast_namespace: Option<&str>,
-        subtopics: Vec<String>,
-        waku_node_key: Option<String>,
-        waku_host: Option<String>,
-        waku_port: Option<String>,
-        waku_addr: Option<String>,
-    ) -> Result<GraphcastAgent, GraphcastAgentError> {
+
+    pub async fn new(config: GraphcastAgentConfig) -> Result<GraphcastAgent, GraphcastAgentError> {
```

Comment:
Can refactor a little bit to make the function body more concise and easier to see what parameters are expected by the function signature
```
pub async fn new(GraphcastAgentConfig {
            wallet_key,
            radio_name,
            registry_subgraph,
            network_subgraph,
            graph_node_endpoint,
            boot_node_addresses,
            graphcast_namespace,
            subtopics,
            waku_node_key,
            waku_host,
            waku_port,
            waku_addr,
        }: GraphcastAgentConfig) -> Result<GraphcastAgent, GraphcastAgentError> {
```

Reply:
ooh I didn't know destructuring can happen like that, nice!

Reply:
come to think of it, since we already have `GraphcastAgentConfig` as the input type for the `config` var, why do you think this is a better structure? 🤔 especially since it would be nice to just have the `config` to later on do `config.validate_set_up()` inside the function. Also I was playing around with gpt4 on this and here's a snippet of why the original is more idiomatic Rust according to it:
```
The reason why this is considered more idiomatic in Rust is mainly due to the following reasons:

1. **Clearer function signature**: By passing the entire `GraphcastAgentConfig` struct as a single parameter, you make the function signature more concise and easier to read. This also makes it clear that the function depends on all the fields of the `GraphcastAgentConfig` struct.

2. **Easier to call**: When you need to call the `new` function, you only need to provide a single `GraphcastAgentConfig` instance as an argument. This simplifies the function call, as you don't need to list all the fields of the struct individually. This is particularly useful when the struct has many fields, as it can become unwieldy to pass all of them one by one.

3. **Separation of concerns**: By destructuring the `GraphcastAgentConfig` struct inside the function body, you separate the function signature (i.e., what inputs it takes and what output it returns) from the details of how the input is used within the function. This makes it easier to understand the function's purpose at a glance and allows you to change the internal implementation without affecting the function's external API.

4. **Extensibility**: In case the `GraphcastAgentConfig` struct is extended with new fields in the future, you won't need to modify the function signature. This can help prevent breaking changes in the API and minimize the impact of such changes on the code that calls the function.

Overall, using a single parameter for the entire `GraphcastAgentConfig` struct makes the function signature cleaner, easier to use, and more maintainable in the long run. This approach aligns with the Rust community's best practices and promotes code readability and simplicity.

```